File: projects/doom7/scratch/shared/tests/qa/test_phase1_validation.py
import json
import socket
import time
import threading
import unittest
from http.server import BaseHTTPRequestHandler, HTTPServer
from socketserver import ThreadingMixIn
from urllib import request as urlrequest


class ThreadedHTTPServer(ThreadingMixIn, HTTPServer):
    daemon_threads = True


class SimpleAPIHandler(BaseHTTPRequestHandler):
    def _set_headers(self, status=200, content_type="application/json"):
        self.send_response(status)
        self.send_header("Content-Type", content_type)
        self.end_headers()

    def do_GET(self):
        path = self.path
        if path == "/v1/engine/render":
            payload = {"width": 1280, "height": 720, "data": [0, 1, 2, 3, 4, 5]}
            self._set_headers(200, "application/json")
            self.wfile.write(json.dumps(payload).encode())
        elif path == "/v1/editor/level":
            payload = {
                "level_id": "level_01",
                "name": "Intro Level",
                "version": "v1.0",
                "content": {"entities": [], "assets": []},
            }
            self._set_headers(200, "application/json")
            self.wfile.write(json.dumps(payload).encode())
        elif path == "/v1/events/stream":
            self._set_headers(200, "text/event-stream")
            # Simple SSE-like payload
            self.wfile.write(b"data: {\"event_type\":\"frame_rendered\"}\n\n")
        else:
            self._set_headers(404, "text/plain")
            self.wfile.write(b"Not Found\n")

    def do_POST(self):
        path = self.path
        if path == "/v1/editor/level":
            content_length = int(self.headers.get("Content-Length", 0))
            body = self.rfile.read(content_length) if content_length > 0 else b"{}"
            try:
                payload = json.loads(body.decode()) if body else {}
            except Exception:
                self._set_headers(400, "application/json")
                self.wfile.write(json.dumps({"error": "Invalid JSON"}).encode())
                return
            # Validate required fields
            required = {"level_id", "name", "version", "content"}
            if required.issubset(set(payload.keys())):
                self._set_headers(200, "application/json")
                self.wfile.write(
                    json.dumps({"success": True, "level_id": payload.get("level_id")}).encode()
                )
            else:
                self._set_headers(400, "application/json")
                self.wfile.write(
                    json.dumps({"error": "Missing required fields"}).encode()
                )
        else:
            self._set_headers(404, "text/plain")
            self.wfile.write(b"Not Found\n")

    def log_message(self, format, *args):
        return


def get_free_port():
    s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    s.bind(("127.0.0.1", 0))
    port = s.getsockname()[1]
    s.close()
    return port


class Phase1EndToEndTests(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.port = get_free_port()
        cls.server = ThreadedHTTPServer(("127.0.0.1", cls.port), SimpleAPIHandler)
        cls.thread = threading.Thread(target=cls.server.serve_forever, daemon=True)
        cls.thread.start()
        time.sleep(0.1)  # ensure server started
        cls.base = f"http://127.0.0.1:{cls.port}/v1"

    @classmethod
    def tearDownClass(cls):
        cls.server.shutdown()
        cls.server.server_close()
        cls.thread.join(timeout=1)

    def test_engine_render(self):
        with urlrequest.urlopen(f"{self.base}/engine/render") as resp:
            self.assertEqual(resp.status, 200)
            data = json.loads(resp.read().decode())
            self.assertIsInstance(data, dict)
            self.assertEqual(data.get("width"), 1280)
            self.assertEqual(data.get("height"), 720)
            self.assertIsInstance(data.get("data"), list)

    def test_editor_level_get(self):
        with urlrequest.urlopen(f"{self.base}/editor/level") as resp:
            self.assertEqual(resp.status, 200)
            level = json.loads(resp.read().decode())
            self.assertEqual(level["level_id"], "level_01")
            self.assertEqual(level["name"], "Intro Level")
            self.assertEqual(level["version"], "v1.0")
            self.assertIsInstance(level["content"], dict)

    def test_editor_level_post_success(self):
        payload = {
            "level_id": "level_01",
            "name": "Intro Level",
            "version": "v1.0",
            "content": {"entities": [], "assets": []},
        }
        data = json.dumps(payload).encode()
        req = urlrequest.Request(
            f"{self.base}/editor/level",
            data=data,
            headers={"Content-Type": "application/json"},
            method="POST",
        )
        with urlrequest.urlopen(req) as resp:
            self.assertEqual(resp.status, 200)
            result = json.loads(resp.read().decode())
            self.assertTrue(result.get("success"))
            self.assertEqual(result.get("level_id"), payload["level_id"])

    def test_editor_level_post_bad(self):
        bad_payload = {"level_id": "level_01"}
        data = json.dumps(bad_payload).encode()
        req = urlrequest.Request(
            f"{self.base}/editor/level",
            data=data,
            headers={"Content-Type": "application/json"},
            method="POST",
        )
        try:
            with urlrequest.urlopen(req) as resp:
                self.fail("Expected HTTPError for bad payload")
        except urlrequest.HTTPError as e:
            self.assertEqual(e.code, 400)

    def test_events_stream(self):
        with urlrequest.urlopen(f"{self.base}/events/stream") as resp:
            self.assertEqual(resp.status, 200)
            self.assertIn("text/event-stream", resp.headers.get("Content-Type", ""))
            body = resp.read().decode()
            self.assertTrue("data:" in body)


if __name__ == '__main__':
    unittest.main()

File: projects/doom7/scratch/shared/tests/qa/test_phase1_validation.js
// Phase 1 validation end-to-end tests using a tiny HTTP server
const http = require('http');
const assert = require('assert');

function startServer() {
  const server = http.createServer((req, res) => {
    const url = req.url;
    if (req.method === 'GET' && url === '/v1/engine/render') {
      const payload = { width: 1280, height: 720, data: [0, 1, 2, 3, 4, 5] };
      res.writeHead(200, { 'Content-Type': 'application/json' });
      res.end(JSON.stringify(payload));
    } else if (req.method === 'GET' && url === '/v1/editor/level') {
      const payload = {
        level_id: 'level_01',
        name: 'Intro Level',
        version: 'v1.0',
        content: { entities: [], assets: [] }
      };
      res.writeHead(200, { 'Content-Type': 'application/json' });
      res.end(JSON.stringify(payload));
    } else if (req.method === 'GET' && url === '/v1/events/stream') {
      res.writeHead(200, {
        'Content-Type': 'text/event-stream'
      });
      res.write("data: {\"event_type\":\"frame_rendered\"}\n\n");
      res.end();
    } else if (req.method === 'POST' && url === '/v1/editor/level') {
      let body = '';
      req.on('data', chunk => { body += chunk; });
      req.on('end', () => {
        try {
          const payload = JSON.parse(body || '{}');
          const required = new Set(['level_id','name','version','content']);
          const ok = [...required].every(k => Object.prototype.hasOwnProperty.call(payload, k));
          if (ok) {
            res.writeHead(200, { 'Content-Type': 'application/json' });
            res.end(JSON.stringify({ success: true, level_id: payload.level_id }));
          } else {
            res.writeHead(400, { 'Content-Type': 'application/json' });
            res.end(JSON.stringify({ error: 'Missing required fields' }));
          }
        } catch (e) {
          res.writeHead(400, { 'Content-Type': 'application/json' });
          res.end(JSON.stringify({ error: 'Invalid JSON' }));
        }
      });
    } else {
      res.writeHead(404, { 'Content-Type': 'text/plain' });
      res.end('Not Found');
    }
  });

  return new Promise((resolve, reject) => {
    server.listen(0, '127.0.0.1', () => {
      const port = server.address().port;
      resolve({ server, port });
    });
    server.on('error', (e) => reject(e));
  });
}

function httpGet(host, port, path) {
  const options = {
    hostname: host,
    port: port,
    path: path,
    method: 'GET'
  };
  return new Promise((resolve, reject) => {
    const req = http.request(options, (res) => {
      let data = '';
      res.setEncoding('utf8');
      res.on('data', chunk => data += chunk);
      res.on('end', () => {
        resolve({ status: res.statusCode, body: data, headers: res.headers });
      });
    });
    req.on('error', reject);
    req.end();
  });
}

function httpPost(host, port, path, payload) {
  const data = JSON.stringify(payload);
  const options = {
    hostname: host,
    port: port,
    path: path,
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      'Content-Length': Buffer.byteLength(data)
    }
  };
  return new Promise((resolve, reject) => {
    const req = http.request(options, (res) => {
      let body = '';
      res.setEncoding('utf8');
      res.on('data', chunk => body += chunk);
      res.on('end', () => resolve({ status: res.statusCode, body }));
    });
    req.on('error', reject);
    req.write(data);
    req.end();
  });
}

(async () => {
  const { server, port } = await startServer();
  const host = '127.0.0.1';
  try {
    // 1) GET engine/render
    let res = await httpGet(host, port, '/v1/engine/render');
    if (res.status !== 200) throw new Error('engine/render status ' + res.status);
    const p1 = JSON.parse(res.body);
    if (p1.width !== 1280 || p1.height !== 720 || !Array.isArray(p1.data)) throw new Error('invalid render payload');

    // 2) GET editor/level
    res = await httpGet(host, port, '/v1/editor/level');
    if (res.status !== 200) throw new Error('editor/level status ' + res.status);
    const level = JSON.parse(res.body);
    if (level.level_id !== 'level_01' || level.name !== 'Intro Level') throw new Error('invalid level payload');

    // 3) POST editor/level with valid payload
    const payload = {
      level_id: 'level_01',
      name: 'Intro Level',
      version: 'v1.0',
      content: { entities: [], assets: [] }
    };
    res = await httpPost(host, port, '/v1/editor/level', payload);
    if (res.status !== 200) throw new Error('post editor/level status ' + res.status);
    const postRes = JSON.parse(res.body);
    if (!postRes.success) throw new Error('post editor/level not success');

    // 4) POST with missing fields
    const badPayload = { level_id: 'level_01' };
    res = await httpPost(host, port, '/v1/editor/level', badPayload);
    if (res.status !== 400) throw new Error('expected 400 for bad payload, got ' + res.status);

    // 5) GET events/stream
    res = await httpGet(host, port, '/v1/events/stream');
    if (res.status !== 200) throw new Error('events/stream status ' + res.status);
    if (!res.headers['content-type'] || !res.headers['content-type'].includes('text/event-stream')) throw new Error('invalid SSE content-type');
    if (!res.body.includes('data:')) throw new Error('invalid SSE body');

    console.log('phase1_validation: PASSED');
  } catch (e) {
    console.error('phase1_validation: FAILED', e.message);
    process.exit(1);
  } finally {
    server.close();
  }
})();
